File: solutions/python/diamond/1/diamond.py
from itertools import chain


def row(size: int, step: int) -> str:
    letter = chr(ord("A") + step)
    return (
        " " * (size - step)
        + (letter if step == 0 else letter + " " * (2 * step - 1) + letter)
        + " " * (size - step)
    )


def rows(letter: str) -> list[str]:
    size = ord(letter) - ord("A")
    return [
        row(size, i) for i in chain(range(size + 1), range(size - 1, -1, -1))
    ]

File: solutions/go/diamond/2/diamond.go
package diamond

import (
	"errors"
	"strings"
)

func row(size, step byte) string {
	letter := 'A' + step
	b := strings.Builder{}
	for range size - step {
		b.WriteRune(' ')
	}
	if step == 0 {
		b.WriteByte(letter)
	} else {
		b.WriteByte(letter)
		for range 2*step - 1 {
			b.WriteRune(' ')
		}
		b.WriteByte(letter)
	}
	for range size - step {
		b.WriteRune(' ')
	}
	return b.String()
}

func Gen(char byte) (string, error) {
	if char < 'A' || char > 'Z' {
		return "", errors.New("char must be in range [A-Z]")
	}
	size := char - 'A'
	var rows []string
	for i := range size + 1 {
		rows = append(rows, row(size, i))
	}
	for i := int(size) - 1; i >= 0; i-- {
		rows = append(rows, row(size, byte(i)))
	}
	return strings.Join(rows, "\n"), nil
}
